### ISSUE TITLE
Allow a way to prevent clangd from auto completing symbols that cannot be 'overtyped'

### ISSUE THREAD

**Issue by AWH1122** [clangd/clangd#2277]
When using clangd in VSCode, I often encounter an issue where, after accepting a completion, the symbol at the end cannot be overtyped, resulting in me typing the symbol twice.

It would be ideal to allow overtyping for symbols like <btn>><\btn> and <btn>;<\btn>, but I’m unsure whether VSCode supports this.

An alternative solution might be implementating a flag that allows for the disabling the suggestion of these final symbols in completions altogether.

**Comment by HighCommander4** [comment]
Can you give an example please? I'm not sure I follow the suggestion based on the description so far.

**Comment by AWH1122** [comment]
https://github.com/user-attachments/assets/3d332599-79ae-486c-a928-916bd7a7cd2e

On this screen recording I accept a tab completion that includes an closing angle bracket. When trying to overtype the symbol the symbol just gets typed twice. This is unlike with symbols like parenthesise which will not be typed twice and the cursor will just jump to the position after that symbol. This can be seen below:

https://github.com/user-attachments/assets/37f8d583-f033-4bcb-b7ef-6fb20b17c5ae


I am also aware that since <kbd>>></kbd> is an operator it may be hard to allow overtyping in this context. Just having an option to prevent completions suggesting the closing symbol would work nicely in lieu.

**Comment by HighCommander4** [comment]
Thanks, I understand now.

For completeness, though your second example doesn't involve accepting a completion proposal, the same behaviour is seen when completion is involved. For example:

```c++
void foo(int);

int main() {
  fo^
}
```

If you invoke code completion and select the proposal for `foo(int)`, you now get:

```c++
void foo(int);

int main() {
  foo(^)
}
```

If you now type an argument followed by `')`, the `')` that you types replaces ("overtypes", as you say) the `)` character that's already in the source.

Whereas, in a similar example with angle brackets:

```c++
template <typename>
class vector {};

int main() {
  vec^
}
```

Accepting the proposal we get:

```c++
template <typename>
class vector {};

int main() {
  vector<^>
}
```

but if you now type `int>`, you get:

```
template <typename>
class vector {};

int main() {
  vector<int>>
}
```

---

As far as I can tell, clangd is not doing anything differently between the two cases. The editor is choosing to apply its "overyping" heuristic to the `()` case but no the `<>` case.

So maybe the most promising path forward would be to file an issue at https://github.com/microsoft/vscode/issues/ suggesting that the editor also apply this behaviour to `<>`?

**Comment by HighCommander4** [comment]
> An alternative solution might be implementating a flag that allows for the disabling the suggestion of these final symbols in completions altogether.

As it happens, the forthcoming clangd 20 aready has such a flag. It can be specified in a `.clangd` file like this:

```yaml
Completion:
  ArgumentLists: OpenLimiter
```

With this setting, upon accepting the proposal in the `vector` case, you get:

```c++
template <typename>
class vector {};

int main() {
  vector<^
}
```

and in the function case, you get:

```c++
void foo(int);

int main() {
  foo(^
}
```

However, unlike `vector<>`, the `#include <>` completion does not currently respect the `OpenDelimiters` setting. That would be an actionable change to make on the clangd side, to make it respect it.

If you try the `OpenDelimiters` setting and find other completions besides `#include` that should respect the setting but do not, please share them.

**Comment by AWH1122** [comment]
I will look at filing an issue on the vscode repository. Good to hear that the next major release of clangd will include this functionality.

> However, unlike vector<>, the #include <> completion does not currently respect the OpenDelimiter setting. That would be an actionable change to make on the clangd side, to make it respect it.

Do you think this behaviour will be added before release, or is it too soon to say?

**Comment by HighCommander4** [comment]
> > However, unlike vector<>, the #include <> completion does not currently respect the OpenDelimiter setting. That would be an actionable change to make on the clangd side, to make it respect it.
> 
> Do you think this behaviour will be added before release, or is it too soon to say?

I'm unlikely to work on this before the release personally; I have a long backlog of other tasks I plan to work on.

I can't speak for anyone else. Clangd is an open-source project, and anyone can submit a fix or enhancement.

**Comment by JVApen** [comment]
@HighCommander4 I've seen variations of this occuring when trying to use the autocomplete to add the required includes.
````
auto f()
{
    return g^(1, 2);
}
````
Became
````
#include "g.h"
auto f()
{
    return g()(1, 2);
}
````
And
````
auto f()
{
    return g^<int, int>();
}
````
Became
````
#include "g.h"
auto f()
{
    return g<T1, T2><int, int>();
}
````

Do you think all these can be solved by VS Code?

**Comment by HighCommander4** [comment]
> Do you think all these can be solved by VS Code?

No, the cases discussed in the previous comment would need server support. It would basically involve checking if the text after the cursor position looks like there's already an argument list there, and if so then skip inserting the argument list (even if the `ArgumentList` config option otherwise specifies than an argument list should be inserted).

This is basically #119, which was duped over to #387, so it might make sense to post these examples in #387.
